jax/version.py

```diff
@@ -14,18 +14,101 @@
 
 # This file is included as part of both jax and jaxlib. It is also
 # eval()-ed by setup.py, so it should not have any dependencies.
+from __future__ import annotations
 
-import os
 import datetime
+import os
+import subprocess
 
 _version = "0.4.16"
+# The following is overwritten by build scripts in releases & distributions.
+# Do not modify this manually, or jax/jaxlib build will fail.
+_release_version: str | None = None
+
+
+def _get_version_string() -> str:
+  if _release_version is not None:
+    return _release_version
+  try:
+    return _version_from_git_tree(_version)
+  except:
+    return _version_from_todays_date(_version)
+
+
+def _version_from_todays_date(base_version: str) -> str:
+  datestring = datetime.date.today().strftime("%Y%m%d")
+  return f"{base_version}.dev{datestring}"
+
+
+def _version_from_git_tree(base_version: str) -> str:
+  stdout = subprocess.check_output(["git", "show", "-s", "--format=%at", "HEAD"])
+  timestamp = int(stdout.decode().strip())
+  datestring = datetime.date.fromtimestamp(timestamp).strftime("%Y%m%d")
+  assert datestring.isnumeric()
+
+  stdout = subprocess.check_output(["git", "describe", "--long", "--always"])
+  commit_hash = stdout.decode().strip().rsplit('-', 1)[-1]
+  assert commit_hash.isalnum()
+
+  return f"{base_version}.dev{datestring}+{commit_hash}"
+
+
+def get_version_for_build() -> str:
+  """Determine the version at build time.
+
+  The returned version string depends on which environment variables are set:
+  - if JAX_RELEASE is set: version looks like "0.4.16"
+  - if JAX_NIGHTLY is set: version looks like "0.4.16.dev20230906"
+  - if neither is set: version looks like "0.4.16.dev20230906+ge58560fdc
+  """
+  if _release_version is not None:
+    return _release_version
+  if os.environ.get('JAX_NIGHTLY'):
+    return _version_from_todays_date(_version)
+  if os.environ.get('JAX_RELEASE') or os.environ.get('JAXLIB_RELEASE'):
+    return _version
+  try:
+    return _version_from_git_tree(_version)
+  except:
+    # Fallback to date string if git is not available.
+    return _version_from_todays_date(_version)
+
+
+def _write_version(fname: str) -> None:
+  """Used by setup.py to write the specified version info into the source tree."""
+  release_version = get_version_for_build()
+  old_version_string = "release_version: str | None = None"
+  new_version_string = f"release_version: str = {str(release_version)!r}"
+  with open(fname) as f:
```

Comment:
Note I have a slight preference for using pathlib but not a strong one. It's much cleaner!

```
fname = pathlib.Path(fname)
contents = fname.read_text()
fname.write_text(new_contents)
```

Reply:
Good idea! Done